packages/byods/src/BaseClient.ts

```diff
@@ -0,0 +1,184 @@
+import fetch, {Response, RequestInit} from 'node-fetch';
+import TokenManager from './TokenManager';
+import DataSourceClient from './DataSourceClient';
+import {ApiResponse, HttpClient} from './types';
+
+export default class BaseClient {
+  private baseUrl: string;
+  private headers: Record<string, string>;
+  private tokenManager: TokenManager;
+  private orgId: string;
+
+  public dataSource: DataSourceClient;
+
+  /**
+   * Creates an instance of BaseClient.
+   * @param {string} baseUrl - The base URL for the API.
+   * @param {Record<string, string>} headers - The headers to be used in requests.
+   * @param {TokenManager} tokenManager - The token manager instance.
+   * @param {string} orgId - The organization ID.
+   */
+  constructor(
+    baseUrl: string,
+    headers: Record<string, string>,
+    tokenManager: TokenManager,
+    orgId: string
+  ) {
+    this.baseUrl = baseUrl;
+    this.headers = headers;
+    this.tokenManager = tokenManager;
+    this.orgId = orgId;
+    this.dataSource = new DataSourceClient(this.getHttpClientForOrg());
+  }
+
+  /**
+   * Retrieves the access token.
+   * @private
+   * @returns {Promise<string>} - The access token.
+   */
+  private async getToken(): Promise<string> {
+    const serviceAppAuthorization = await this.tokenManager.getServiceAppAuthorization(this.orgId);
+    let token = serviceAppAuthorization.token.accessToken;
+
+    if (new Date() >= new Date(serviceAppAuthorization.token.expiresAt)) {
+      await this.tokenManager.refreshServiceAppAccessToken(this.orgId, this.headers);
+      const refreshedAuthorization = await this.tokenManager.getServiceAppAuthorization(this.orgId);
+      token = refreshedAuthorization.token.accessToken;
+    }
+    // TODO: Handle refresh token expiration
+
+    return token;
+  }
+
+  /**
+   * Makes an HTTP request.
+   * @param {string} endpoint - The API endpoint.
+   * @param {RequestInit} [options={}] - The request options.
+   * @returns {Promise<ApiResponse<T>>} - The API response.
+   * @template T
+   */
+  public async request<T>(endpoint: string, options: RequestInit = {}): Promise<ApiResponse<T>> {
+    const url = `${this.baseUrl}${endpoint}`;
+    const token = await this.getToken();
+
+    const response: Response = await fetch(url, {
+      ...options,
+      headers: {
+        Authorization: `Bearer ${token}`,
+        ...this.headers,
+        ...options.headers,
+      },
+    });
+
+    const data: any = await response.json();
+    if (!response.ok) {
+      throw new Error(`Error: ${response.status} - ${data.message}`);
+    }
+
+    return {data, status: response.status};
+  }
+
+  /**
+   * Makes a POST request.
+   * @param {string} endpoint - The API endpoint.
+   * @param {Record<string, any>} body - The request body.
+   * @param {Record<string, string>} [headers={}] - The request headers.
+   * @returns {Promise<ApiResponse<T>>} - The API response.
+   * @template T
+   */
+  public async post<T>(
```

Comment:
Let's move these methods to a http utility and use accordingly as discussed.

Reply:
This seems to be a bigger effort. As discussed 1:1 , Created SPARK-563120 to handle this.